text/0156-columnar-encoding.md

```diff
@@ -0,0 +1,411 @@
+    // Time range
+    Field::new("start_time_unix_nano", DataType::UInt64, false),
+    Field::new("end_time_unix_nano", DataType::UInt64, false),
+
+    Field::new("trace_id", DataType::Binary, false),
+    Field::new("span_id", DataType::Binary, false),
+    Field::new("trace_state", DataType::Utf8, true),
+    Field::new("parent_span_id", DataType::Binary, true),
+    Field::new("name", DataType::Utf8, false),
+    Field::new("kind", DataType::UInt8, false),
+
+    // Attributes
+    Field::new("attributes", DataType::List(
+        Box::new(Field::new(
+            "attribute",
+            DataType::Struct(vec![
+                Field::new("name", DataType::Utf8, false),
+                Field::new("value", DataType::Utf8, false),
+            ]),
+            true,
+        ))
+    ), true),
+
+    // Events
+    Field::new("events", DataType::List(
+        Box::new(Field::new(
+            "event",
+            DataType::Struct(vec![
+                Field::new("time_unix_nano", DataType::UInt64, false),
+                Field::new("name", DataType::Utf8, false),
+                Field::new("attributes", DataType::List(
+                    Box::new(Field::new(
+                        "attribute",
+                        DataType::Struct(vec![
+                            Field::new("name", DataType::Utf8, false),
+                            Field::new("value", DataType::Utf8, false),
+                        ]),
+                        true,
+                    ))
+                ), true),
+            ]),
+            true,
+        ))
+    ), true),
+
+    // Links
+    Field::new("links", DataType::List(
+        Box::new(Field::new(
+            "link",
+            DataType::Struct(vec![
+                Field::new("trace_id", DataType::Binary, false),
+                Field::new("span_id", DataType::Binary, false),
+                Field::new("trace_state", DataType::Utf8, false),
+                Field::new("attributes", DataType::List(
+                    Box::new(Field::new(
+                        "attribute",
+                        DataType::Struct(vec![
+                            Field::new("name", DataType::Utf8, false),
+                            Field::new("value", DataType::Utf8, false),
+                        ]),
+                        true,
+                    ))
+                ), true),
+            ]),
+            true,
+        ))
+    ), true),
+])
+```
+
+### Corners cases
+
+Backends that don't support natively multivariate time-series can still automatically transform these events in multiple univariate time-series and operate as usual.
+
+Specialized processors can be developed to group metrics, logs and traces in optimized batch of events in order to connect existing OTEL collectors with backends supporting this new protocol extension.
+
+Specialized processors can be developed to convert batch of events into the existing entities for backends that don't support this protocol extension.
```

Comment:
How this protocol impacts the OpenTelemetry collector will be a very important question. I can easily see a OTel-Columnar receiver that translates into univariate metric streams, but I'm not sure anyone really wants to do that. Most likely, a system that chooses multivariate metrics wants to pass them through the collector in that format. This means the collector would need a new kind of pipeline and processors would modify a new kind of data structure. Questions of this nature will need to be addressed before a protocol can move forward -- what are we committing to for the collector?

Reply:
@jmacd - I believe both will be needed. You're correct that in an ideal world the data would flow through the collector in it's native format; however, there will be a need, hopefully short/medium-term, to translate multi-variate metrics to univariate for the observability systems that do not currently support multi-variate metrics. This would allow products/systems needing multivariate support to be instrumented correctly (without it the data wouldn't be complete and/or couldn't leave the system in an efficient enough manner) and leave translation to downstream systems.

Reply:
This new [diagram](https://github.com/open-telemetry/oteps/pull/171#issuecomment-922156998) provides more details on this topic. I will integrate it into the OTEP document if the reviewers agree with the approach.